dynamic_validators/insecure_dynamic_validators.md

```diff
@@ -0,0 +1,103 @@
+
+
+### Storing dynamically added validators
+As discussed in the previous section it is necessary to (temporarily) store the updates to validators list until `end_block` is executed. Furthermore, in case of a node crash it is required that that upon restart the node should be able to recover its dynamically added validators. Following is one possible solution,
+
+- Store the validators in a MongoDB collection along with a flag i.e.,
+```json
+{"validators": [],
+ "sync": true
+}
+```
+- The sync flag is used during `end_block` to decide if a validator update should be returned.
+- The sync flag is set to `True` when api `PUT /api/v1/validators` is called and the `"validators"` are updated.
+- Whenever the sync flag is set, `end_block` will call `http://tendermint:46657/validators`, calculate the diff with the list of validators retrieved from mongodb and return it.
+- When bigchaindb starts the sync flag should be set to `True`.
+
+
+### API impact
+New api endpoint `PUT /api/v1/validators` will introduced,
+
+- Add/remove validators
+  - Uri: `PUT /api/v1/validators`
+  - Body: 
+  ```json
+  [{"pub_key":{"type":"ed25519","data":"4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A"},
+    "power":10},
+   {"pub_key":{"type":"ed25519","data":"608D839D7100466D6BA6BE79C320F8B81DE93CFAA58CF9768CF921C6371F2553"},
+   "power":0}
+   ]
+   ```
+   - response: `200`
+
+- List validators
+  - Uri: `GET /api/v1/validators`
+  - response: 
+  ```json
+  [{"pub_key":{"type":"ed25519","data":"4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A"},
+    "power":10},
+   {"pub_key":{"type":"ed25519","data":"608D839D7100466D6BA6BE79C320F8B81DE93CFAA58CF9768CF921C6371F2553"},
+   "power":0}
+   ]
+  ```
+
+**NOTE**: when `power: 0` is set for a given validator it is removed from the validators list.
+
+### Security impact
+The current implementation is insecure i.e. any user who gains access to a given node can modify its validators. This node could effectively be considered byzantine and the network can tolerate (<1/3) byzantine nodes.
+
+### Performance impact
+N/A
+
+### End user impact
+N/A
+
+### Deployment impact
+N/A
+
+### Documentation impact
+The new api introduced should be documented along with its current state of security.
+
+
+### Testing impact
+Following test cases should be included:
+- Add a new validator
+- Remove a validator
+- Change the power of a validator
+**NOTE**: The above tests should check if the MongoDB collection is updated and calling `end_block` returns a valid diff. More accurate tests could be written in the integration test suite wherein `http://tendermint:46657/validators` is called to verify the validator changes.
+
+## Implementation
+
+### Assignee(s)
+Primary assignee(s): @kansi
+
+### Targeted Release
+BigchainDB 2.0
```

Comment:
Earlier, you used the terminology "bigchaindb tendermint MVP." That's probably better, because I'm not sure what the final name or version number will be.